File: csharp/step3_env.cs
using System;
using System.Collections.Generic;

namespace Mal
{
    internal class Step3
    {
        Value READ(string arg)
        {
            reader.Set(arg);
            return reader.Read_form();
        }

        Value EVAL(Value arg, Env env)
        {
            if (arg is List lst)
            {
                if (lst.IsEmpty())
                    return arg;

                Value val = lst.First();

                if (val == Reader.Def)
                {
                    if (lst.Count() != 3)
                        throw new MalException(Reader.Def.Name + " - syntax error");

                    val = lst.Nth(1);

                    if (val is Symbol sym)
                    {
                        val = EVAL(lst.Nth(2), env);
                        env.Set(sym, val);

                        return val;
                    }
                    else
                        throw new MalException(Reader.Def.Name + " - symbol expected");
                }
                else if (val == Reader.Let)
                {
                    if (lst.Count() != 3)
                        throw new MalException(Reader.Let.Name + " - syntax error");

                    Env newEnv = new Env(env);

                    Value val_1 = lst.Nth(1);
                    if (val_1 is Sequence syms && syms.Count() % 2 == 0)
                    {
                        int cnt = syms.Count();
                        for (int i = 0; i < cnt; i += 2)
                        {
                            if (syms.Nth(i) is Symbol sym)
                            {
                                newEnv.Set(sym, EVAL(syms.Nth(i + 1), newEnv));
                            }
                            else
                                throw new MalException(Reader.Let.Name + " - bindings - symbol is expected");
                        }

                        return EVAL(lst.Nth(2), newEnv);
                    }
                    else
                        throw new MalException(Reader.Let.Name + " - bad bindings");
                }
                else
                {
                    Value ast = Eval_ast(arg, env);

                    val = (ast as List).First();

                    if (val is Func_Native fn)
                    {
                        return fn.Apply((ast as List).Rest());
                    }
                    else
                        throw new MalException("function is expected");
                }
            }
            else
                return Eval_ast(arg, env);
        }

        string PRINT(Value arg)
        {
            return Printer.Pr_str(arg, true);
        }

        string Rep(string arg, Env env)
        {
            return PRINT(EVAL(READ(arg), env));
        }

        Value Eval_ast(Value arg, Env env)
        {
            if (env == null)
                return arg;

            if (arg == null)
                return arg;

            if (arg is Symbol sym)
            {
                Value val = env.Get(sym);

                if (val == null)
                    throw new MalException(string.Format("'{0}' not found", sym.Name));

                return val;
            }
            else if (arg is List lst)
            {
                if (lst.Elements == null)
                    return arg;

                List<Value> elms = new List<Value>();
                for (List.Node elm = lst.Elements; elm != null; elm = elm.cdr)
                {
                    elms.Add(EVAL(elm.car, env));
                }

                return new List(elms);
            }
            else if (arg is Vector vec)
            {
                List<Value> elms = new List<Value>();

                foreach (Value v in vec.Elements)
                    elms.Add(EVAL(v, env));

                return new Vector(elms);
            }
            else if (arg is HashMap map)
            {
                HashMap hm = new HashMap();
                Dictionary<Value, Value> elms = hm.Elements;

                foreach (var kv in map.Elements)
                    elms[kv.Key] = EVAL(kv.Value, env);

                return hm;
            }

            return arg;
        }

        Env CreateTopLevelEnv()
        {
            Env env = new Env();

            env.Set(Reader.AddSymbol("+") as Symbol, new Func_Add());
            env.Set(Reader.AddSymbol("-") as Symbol, new Func_Sub());
            env.Set(Reader.AddSymbol("*") as Symbol, new Func_Mul());
            env.Set(Reader.AddSymbol("/") as Symbol, new Func_Div());

            return env;
        }

        internal void Repl()
        {
            Env env = CreateTopLevelEnv();

            for (; ;)
            {
                try
                {
                    Console.Write("user> ");
                    var str = Console.ReadLine();
                    if (str == null)
                        break;
                    var val = Rep(str, env);
                    if (val != null)
                        Console.WriteLine(val);
                }
                catch (MalException ex)
                {
                    Console.WriteLine("ERROR: {0}", ex.Message);
                }
            }
        }

        private readonly Reader reader = new Reader("");
    }
}

File: cxx/step3_env.cxx
#include <iostream>

#include "types.h"
#include "reader.h"
#include "printer.h"
#include "core.h"
#include "env.h"
#include "gc.h"

namespace MAL
{

Value * READ(const std::string & arg)
{
   reader rdr(arg);

   return rdr.read_form();
}

Value * eval_ast(Value * arg, EnvPtr & env)
{
   if (arg == nullptr)
      return nullptr;

   if (arg->get_type() == Type::Symbol)
   {
      return env->get((Symbol *)arg);
   }

   if (arg->get_type() == Type::List)
   {
      return ((List *) arg)->map([&](Value * elm)
                                 { return EVAL(elm, env); });
   }

   if (arg->get_type() == Type::Vector)
   {
      return ((Vector *) arg)->map([&](Value * elm)
                                   { return EVAL(elm, env); });
   }

   if (arg->get_type() == Type::AVLTree)
   {
      return ((AVLTree *)arg)->map([&](Value * elm)
                                   { return EVAL(elm, env); });
   }

   return arg;
}

Value * EVAL(Value * arg, EnvPtr env)
{
   if (Value::is_type(arg, Type::List))
   {
      List * lst = (List *)arg;

      Value * fst = lst->first();
      if (fst == Symbol::Def)
      {
         if (lst->length() != 3)
            error(0, Symbol::Def, "syntax error");

         Value * sym = lst->nth(1);
         if (!Value::is_type(sym, Type::Symbol))
            error_expected(1, "symbol", Symbol::Def);

         Value * val = EVAL(lst->nth(2), env);

         env->set((Symbol *)sym, val);

         return val;
      }

      if (fst == Symbol::Let)
      {
         if (lst->length() != 3)
            error(0, Symbol::Let, "syntax error");

         Value * vars = lst->nth(1);
         if (!Value::is_sequence(vars))
            error_expected(1, "sequence of variable definition", Symbol::Let);

         Sequence * seq = (Sequence *)vars;
         int count =  seq->length();

         if (count % 2 != 0)
            error(1, Symbol::Let, "variable definition syntax error");

         EnvPtr env2( new Env(env) );

         for (int i=0; i < count; i+=2)
         {
            Value * var = seq->nth(i);
            Value * val = seq->nth(i + 1);

            if (!Value::is_type(var, Type::Symbol))
               error_expected(i+1, "variable", Symbol::Let);

            env2->set((Symbol *)var, EVAL(val, env2));
         }

         return EVAL(lst->nth(2), env2);
      }
   }

   Value * ast = eval_ast(arg, env);

   if (ast != nullptr)
   {
      if (ast->get_type() == Type::List)
      {
         List * lst = (List *)ast;

         if (lst->is_null())
            return lst;

         Value * fst = lst->first();

         if (!Value::is_type(fst, Type::Function))
            error_expected(0, "function", "apply");

         return ((Function *) fst)->eval(lst->elements()->get_cdr());
      }
   }

   return ast;
}

std::string PRINT(Value * arg)
{
   return pr_str(arg, true);
}

std::string rep(const std::string & arg, EnvPtr & env)
{
   return PRINT(EVAL(READ(arg), env));
}

void repl(const char * prompt, EnvPtr & env)
{
   std::string input;

   while (readline(prompt, input))
   {
      try
      {
         std::cout << rep(input, env) << std::endl;
      }
      catch (MalException & ex)
      {
         std::cout << "Error: " << ex.get_msg() << std::endl;
      }
   }
}

void set_top_level_env(EnvPtr & env)
{
   env->set("+",   Function::create("+",   c_add));
   env->set("-",   Function::create("-",   c_sub));
   env->set("*",   Function::create("*",   c_mul));
   env->set("/",   Function::create("/",   c_div));
}

}

int main(int argc, char * argv[])
{
   MAL::GC::initialize(1024*1024);
   MAL::EnvPtr env(new MAL::Env());

   MAL::set_top_level_env(env);
   MAL::repl("user> ", env);

   return 0;
}
